### PR TITLE
CALS-5811: Submit available on RFA01b 

### PR DESCRIPTION
**Pull request by lauherkcambria**
#### Which User story does this relate to (link)?
-https://osi-cwds.atlassian.net/browse/CALS-5811

#### Brief feature description
- add submit validations to B form 

#### Does this PR block/effect other stories?
-n/a

#### Do any commands needs to be run in order to run the project with this PR?
- yarn rebuild-assets-dev

#### Code coverage percent diff as a result of this PR?
-85.44% Statements 
75.04% Branches 
82.01% Functions 
85.84% Lines

### CODE REVIEW

File: app/javascript/rfa_forms/rfa01a_edit_view/index.js [hunk 1]
```diff
@@ -71,11 +71,13 @@ export default class Rfa01EditView extends React.Component {
       const county = this.props.countyTypes.find(countyType => countyType.id === parseInt(countyValueId))
       this.setApplicationState('application_county', county)
     }
+    const DataValidForSubmit = this.validateAllRequiredForSubmit(this.props.application)
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01a_edit_view/index.js]: Now that we have the data field in API, the `componentWillMount` can be removed and status for submit can be calculated in constructor along with the county code
```
  const submitEnabled = this.props.application.metadata && this.props.application.metadata.submit_enabled

    this.state = {
      focusComponentName: '',
      activeNavLinkHref: '',
      application: Immutable.fromJS(this.props.application),
      errors: {},
      disableSubmit: !submitEnabled
    }

    if (!this.state.application.application_county) {
      const countyValue = (this.props.user && this.props.user.county_code)
      this.state.application.application_county = this.props.countyTypes.find(countyType => countyType.id === parseInt(countyValue))
    }
```

File: app/javascript/rfa_forms/rfa01a_edit_view/index.js [hunk 2]
```diff
@@ -71,11 +71,13 @@ export default class Rfa01EditView extends React.Component {
...
+    this.setApplicationState('metadata', {'submit_enabled': !DataValidForSubmit})
   }
   componentDidMount () {
     const DataValidForSubmit = !this.validateAllRequiredForSubmit(this.props.application)
     const DataValidForSave = !checkForNameValidation(this.props.application.applicants)
-
+    this.setApplicationState('metadata', {'submit_enabled': !DataValidForSubmit})
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01a_edit_view/index.js]: After the constructor change, you won't need line 78, 80, 82.

File: app/javascript/rfa_forms/rfa01b_edit_view/criminalFields.js [hunk 3]
```diff
@@ -9,6 +9,44 @@ import {disclosureDefaults} from 'constants/defaultFields'
 import Button from 'components/common/button'
 
 export default class CriminalFields extends React.Component {
+  constructor (props) {
+    super(props)
+
+    this.offenseId = this.props.validatorPrefix + '].offense'
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01b_edit_view/criminalFields.js]: the end bracket and . `].` should be in the parent component.

File: app/javascript/rfa_forms/rfa01b_edit_view/crimeBackgroundAgainstCohabitant.js [hunk 4]
```diff
@@ -64,14 +64,17 @@ export default class CrimeBackgroundAgainstCohabitant extends React.Component {
                   crime={crime}
                   idPrefix='crimeBackgroundAgainstCohabitant'
                   clickClose={this.clickClose}
-                  onFieldChange={this.onFieldChange} />
+                  onFieldChange={this.onFieldChange}
+                  validator={this.props.validator}
+                  validatorPrefix={'arrested_for_crime_disclosures.[' + index}
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01b_edit_view/crimeBackgroundAgainstCohabitant.js]: Any reason to create a separate `validatorPrefix` and not change the value of `idPrefix`?

File: app/javascript/rfa_forms/rfa01b_edit_view/index.js [hunk 5]
```diff
@@ -19,7 +19,7 @@ import Validator from 'helpers/validator'
 import PageTemplate from 'components/common/pageTemplate'
 import {disclosureDefaults} from 'constants/defaultFields'
 
-export default class Rfa01bList extends React.Component {
+export default class Rfa01bEditView extends React.Component {
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01b_edit_view/index.js]: Good catch.

File: app/javascript/rfa_forms/rfa01b_edit_view/index.js [hunk 6]
```diff
@@ -39,15 +39,52 @@ export default class Rfa01bList extends React.Component {
...
+  validateAllRequiredForSubmit (data) {
+    let requiredRules = this.validator.allIsRequiredRules()
+    requiredRules = requiredRules.merge(this.validator.allValidationsWithOnlyRule('isRequiredBoolean'))
+    requiredRules = requiredRules.merge(this.validator.allValidationsWithOnlyRule('isRequiredIf'))
+
+    return this.validator.validateAllFieldsWithRules(data, requiredRules)
+  }
+
+  componentWillMount () {
+    const DataValidForSubmit = this.props.rfa_a01_application.metadata && this.props.rfa_a01_application.metadata.submit_enabled && this.validateAllRequiredForSubmit(this.props.rfa_b01_application)
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01b_edit_view/index.js]: Similar to my comment for `componentWillMount` in rfa01a, you wouldn't need this `componentWillMount`

File: app/javascript/rfa_forms/rfa01b_edit_view/index.js [hunk 7]
```diff
@@ -39,15 +39,52 @@ export default class Rfa01bList extends React.Component {
...
+    let application = this.props.rfa_b01_application
+    application.metadata = {submit_enabled: !DataValidForSubmit}
+    this.setState({application: application})
+  }
+  componentDidMount () {
+    const DataValidForSubmit = !this.validateAllRequiredForSubmit(this.props.rfa_b01_application)
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01b_edit_view/index.js]: Lines related to submit button can be removed once added to constructor.

File: app/javascript/rfa_forms/rfa01a_edit_view/index.js [hunk 8]
```diff
@@ -86,6 +88,7 @@ export default class Rfa01EditView extends React.Component {
 
     if (this.state.disableSubmit !== DataValidForSubmit) {
       this.setState({disableSubmit: DataValidForSubmit})
+      this.setApplicationState('metadata', {'submit_enabled': !DataValidForSubmit})
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01a_edit_view/index.js]: We should get rid of `componentDidMount ` and move the code to `setApplicationState`. Using `componentDidMount` will result in unnecessary execution when focus card or scroll event updates state.

File: app/javascript/rfa_forms/rfa01b_edit_view/index.js [hunk 9]
```diff
@@ -39,15 +39,52 @@ export default class Rfa01bList extends React.Component {
...
+    const DataValidForSave = !checkForNameValidation(this.props.rfa_a01_application.applicants)
+    let application = this.props.rfa_b01_application
+    application.metadata = {submit_enabled: !DataValidForSubmit}
+    this.setState({application: application})
+    this.setState({disableSave: DataValidForSave})
+    this.setState({disableSubmit: DataValidForSubmit})
+  }
+
+  componentDidUpdate (prevProps, prevState) {
+    const DataValidForSubmit = !this.validateAllRequiredForSubmit(this.state.application)
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01b_edit_view/index.js]: `componentDidUpdate` code should be moved to `setApplicationState`, as I mentioned in rfa01a.

File: app/models/helpers/RFA/adoption_history_helper.rb [hunk 10]
```diff
@@ -3,5 +3,12 @@ class Helpers::Rfa::AdoptionHistoryHelper  < Helpers::ModelHelperBase
   def model_class
     Rfa::AdoptionHistory
   end
+def self.parent_path
```
Comment (ratneshraval) [inline app/models/helpers/RFA/adoption_history_helper.rb]: These are already declared in `adoption_history.rb` do you need them here?

Comment (lauherkcambria) [reply app/models/helpers/RFA/adoption_history_helper.rb]: 👍

File: app/javascript/rfa_forms/rfa01b_edit_view/criminalFields.js (hunk 3 above)
Comment (lauherkcambria) [reply app/javascript/rfa_forms/rfa01b_edit_view/criminalFields.js]: sure, i'll make that preference change.

File: app/javascript/rfa_forms/rfa01b_edit_view/crimeBackgroundAgainstCohabitant.js (hunk 4 above)
Comment (lauherkcambria) [reply app/javascript/rfa_forms/rfa01b_edit_view/crimeBackgroundAgainstCohabitant.js]: These are different props used in different ways. the id prefix is used on ids.  the validation prefix is used in order for the lodash _.get() command to get the correct prop stored on the validation data object.

File: app/javascript/rfa_forms/rfa01b_edit_view/index.js (hunk 9 above)
Comment (nasar-cambria) [reply app/javascript/rfa_forms/rfa01b_edit_view/index.js]: @ratneshraval I tried this. When I move this to ```setApplicationState``` its not auto saving any state changes which are happening inside ```rfa01B Form```. Will it not be useful to use ```componentDidUpdate``` when an operation needs to happen after the DOM is updated and the update queue is emptied

Comment (ratneshraval) [reply app/javascript/rfa_forms/rfa01b_edit_view/index.js]: @nasar-cambria I'm not sure if I'm following. Are you saying that the state changes are not available in the next line after you call `setState`?

**Comment by sunilb0575** [comment]
1. Submit button in RFA-01b is enabled even after missing some of the required fields in 01A.
2. County in RFA-01A on initial load is empty, should be default county.

**Comment by lauherkcambria** [comment]
jenkins, restest this please.

**Comment by lauherkcambria** [comment]
jenkins, retest this please.

**Comment by Nasir2712** [comment]
@lauherkcambria did you get a chance to look at bugs reported by Sunil?

**Comment by ratneshraval** [comment]
@lauherkcambria there is still bug with county. The application object is already immutable, so the code to check/set county doesn't work.

**Comment by sunilb0575** [comment]
@lauherkcambria can you please update the title of the PR to match story title?

**Comment by lauherkcambria** [comment]
jenkins, retest this please.

**Comment by lauherkcambria** [comment]
jenkins, retest this please.

**Comment by lauherkcambria** [comment]
jenkins, retest this please.

**Comment by lauherkcambria** [comment]
jenkins, retest this please.

**Pull request merged**